lightningd/channel_control.c

```diff
@@ -128,6 +141,470 @@ void notify_feerate_change(struct lightningd *ld)
+			    "%s.",
+			    info->err_msg,
+			    type_to_string(tmpctx, struct wally_tx,
+			    		   info->final_tx->wtx));
+	}
+	else
+		handle_tx_broadcast(info);
+
+	tal_free(info);
+}
+
+/* Callback for after the splice tx is sent to bitcoind */
+static void send_splice_tx_done(struct bitcoind *bitcoind UNUSED,
+				bool success, const char *msg,
+				struct send_splice_info *info)
+{
+	/* A NULL value of `info->cc` means we got here without user intiation.
+	 * This means we are the ACCEPTER side of the splice */
+	struct lightningd *ld = info->channel->peer->ld;
+	struct bitcoin_outpoint outpoint;
+
+	bitcoin_txid(info->final_tx, &outpoint.txid);
+	outpoint.n = info->output_index;
+
+	if (!success) {
+		info->err_msg = tal_strdup(info, msg);
+		bitcoind_getutxout(ld->topology->bitcoind, &outpoint,
+				   check_utxo_block, info);
+	} else {
+		handle_tx_broadcast(info);
+		tal_free(info);
+	}
+}
+
+/* Where the splice tx gets finally transmitted to the chain */
+static void send_splice_tx(struct channel *channel,
+			   const struct bitcoin_tx *tx,
+			   struct splice_command *cc,
+			   u32 output_index)
+{
+	struct lightningd *ld = channel->peer->ld;
+	u8* tx_bytes = linearize_tx(tmpctx, tx);
+
+	log_debug(channel->log,
+		  "Broadcasting splice tx %s for channel %s.",
+		  tal_hex(tmpctx, tx_bytes),
+		  type_to_string(tmpctx, struct channel_id, &channel->cid));
+
+	struct send_splice_info *info = tal(NULL, struct send_splice_info);
+
+	info->cc = tal_steal(info, cc);
+	info->channel = channel;
+	info->final_tx = tal_steal(info, tx);
+	info->output_index = output_index;
+	info->err_msg = NULL;
+
+	bitcoind_sendrawtx(ld->topology->bitcoind,
+			   cc ? cc->cmd->id : NULL,
+			   tal_hex(tmpctx, tx_bytes),
+			   false,
+			   send_splice_tx_done, info);
+}
+
+/* After user signs PSBT with splice_signed, our node goes through the signing
+ * process (adding it's own signatures and peers' sigs), sending the result to
+ * us here: */
+static void handle_splice_confirmed_signed(struct lightningd *ld,
+					   struct channel *channel,
+					   const u8 *msg)
+{
+	struct splice_command *cc;
+	struct splice_command *n;
+	struct bitcoin_tx *tx;
+	u32 output_index;
+
+	if (!fromwire_channeld_splice_confirmed_signed(tmpctx, msg, &tx, &output_index)) {
+
+		channel_internal_error(channel,
+				       "bad splice_confirmed_init %s",
+				       tal_hex(channel, msg));
+		return;
+	}
+
+	if (channel->state != CHANNELD_NORMAL) {
+		log_debug(channel->log,
+			  "Would broadcast splice, but state %s"
+			  " isn't CHANNELD_NORMAL",
+			  channel_state_name(channel));
+		return;
+	}
+
+	/* First look for a matching user command for this splice */
+	list_for_each_safe(&ld->splice_commands, cc, n, list) {
+		if (channel != cc->channel)
+			continue;
+
+		channel_set_state(channel,
+				  CHANNELD_NORMAL,
+				  CHANNELD_AWAITING_SPLICE,
+				  REASON_USER,
+				  "Broadcasting splice");
+
+		send_splice_tx(channel, tx, cc, output_index);
```

Comment:
nit: don't duplicate code with below; use a variable set to NULL // if not found.